DataFormats/TauReco/interface/PFTau.h

```diff
@@ -244,24 +257,39 @@ class PFTau : public BaseTau {
 
     float signalConeSize_; 
 
-    reco::PFJetRef jetRef_;
+    reco::JetBaseRef jetRef_;
     PFTauTagInfoRef PFTauTagInfoRef_;
-    reco::PFCandidatePtr leadPFChargedHadrCand_;
-    reco::PFCandidatePtr leadPFNeutralCand_;
-    reco::PFCandidatePtr leadPFCand_;
+    reco::CandidatePtr leadChargedHadrCand_;
+    reco::CandidatePtr leadNeutralCand_;
+    reco::CandidatePtr leadCand_;
     reco::TrackRef electronPreIDTrack_;
 
     // Signal candidates
-    std::vector<reco::PFCandidatePtr> selectedSignalPFCands_;
-    std::vector<reco::PFCandidatePtr> selectedSignalPFChargedHadrCands_;
-    std::vector<reco::PFCandidatePtr> selectedSignalPFNeutrHadrCands_;
-    std::vector<reco::PFCandidatePtr> selectedSignalPFGammaCands_;
+    std::vector<reco::CandidatePtr> selectedSignalCands_;
+    std::vector<reco::CandidatePtr> selectedSignalChargedHadrCands_;
+    std::vector<reco::CandidatePtr> selectedSignalNeutrHadrCands_;
+    std::vector<reco::CandidatePtr> selectedSignalGammaCands_;
 
     // Isolation candidates
-    std::vector<reco::PFCandidatePtr> selectedIsolationPFCands_;
-    std::vector<reco::PFCandidatePtr> selectedIsolationPFChargedHadrCands_;
-    std::vector<reco::PFCandidatePtr> selectedIsolationPFNeutrHadrCands_;
-    std::vector<reco::PFCandidatePtr> selectedIsolationPFGammaCands_;
+    std::vector<reco::CandidatePtr> selectedIsolationCands_;
+    std::vector<reco::CandidatePtr> selectedIsolationChargedHadrCands_;
+    std::vector<reco::CandidatePtr> selectedIsolationNeutrHadrCands_;
+    std::vector<reco::CandidatePtr> selectedIsolationGammaCands_;
+
+    // Transient caches for PFCandidate-based accessors
+    edm::AtomicPtrCache<reco::PFCandidatePtr> leadPFChargedHadrCand_;
+    edm::AtomicPtrCache<reco::PFCandidatePtr> leadPFNeutralCand_;
+    edm::AtomicPtrCache<reco::PFCandidatePtr> leadPFCand_;
```

Comment:
Why bother making these three be member data? Why not just do the conversion when the function requesting them is called?